tajo-catalog/tajo-catalog-server/src/main/java/org/apache/tajo/catalog/store/AbstractDBStore.java

```diff
@@ -2172,6 +2178,252 @@ private void setPartitionKeys(int pid, PartitionDescProto.Builder partitionDesc)
     return partitions;
   }
 
+  /**
+   * Check if list of partitios exist on catalog.
+   *
+   *
+   * @param databaseId
+   * @param tableId
+   * @return
+   */
+  public boolean existPartitionsOnCatalog(int databaseId, int tableId) {
+    Connection conn = null;
+    ResultSet res = null;
+    PreparedStatement pstmt = null;
+    boolean result = false;
+
+    try {
+      String sql = "SELECT COUNT(*) CNT FROM "
+        + TB_PARTTIONS +" WHERE " + COL_TABLES_PK + " = ?  ";
+
+      if (LOG.isDebugEnabled()) {
+        LOG.debug(sql);
+      }
+
+      conn = getConnection();
+      pstmt = conn.prepareStatement(sql);
+      pstmt.setInt(1, tableId);
+      res = pstmt.executeQuery();
+
+      if (res.next()) {
+        if (res.getInt("CNT") > 0) {
+          result = true;
+        }
+      }
+    } catch (SQLException se) {
+      throw new TajoInternalError(se);
+    } finally {
+      CatalogUtil.closeQuietly(pstmt, res);
+    }
+    return result;
+  }
+
+  @Override
+  public List<PartitionDescProto> getPartitionsByFilter(PartitionsByFilterProto request) {
+    throw new TajoRuntimeException(new UnsupportedException());
+  }
+
+  @Override
+  public List<PartitionDescProto> getPartitionsByAlgebra(PartitionsByAlgebraProto request) throws
+      UndefinedDatabaseException, UndefinedTableException, UndefinedPartitionMethodException {
+    Connection conn = null;
+    PreparedStatement pstmt = null;
+    ResultSet res = null;
+    int currentIndex = 1;
+    String selectStatement = null;
+
+    List<PartitionDescProto> partitions = TUtil.newList();
+    List<PartitionFilterSet> filterSets = TUtil.newList();
+
+    try {
+      int databaseId = getDatabaseId(request.getDatabaseName());
+      int tableId = getTableId(databaseId, request.getDatabaseName(), request.getTableName());
+      if (!existPartitionMethod(request.getDatabaseName(), request.getTableName())) {
+        throw new UndefinedPartitionMethodException(request.getTableName());
+      }
+
+      if (!existPartitionsOnCatalog(databaseId, tableId)) {
+        throw new PartitionNotFoundException(request.getTableName());
```

Comment:
Could you elaborate the purpose of this exception? Does it means that there is no partition? If so, it does not make sense because it returns a list of partitions. So, it should return an empty list if there is no partitions.